### PR TITLE
WIP: Live API endpoint for querying FIPS data

### PR DESCRIPTION
**Pull request by bcipolli**
This is initial work on the API endpoint for pulling FIPS data. I created a `fips` Django app, with a simple `FipsEntity` model that contains `name`, `type`, and `fips_id`.

I:
- Added `rest_framework` to `settings.INSTALLED_APPS` (the `/search` API endpoint was returning 500 without it)
- Added the admin interface, added a few real fips entities (`San Fransisco`, `Oakland`, `San Diego`), and got the view to properly filter based on front-end requests.
- Dumped the data as a fixture, made a command for loading all fixtures, and added this command to  the installation section of `README.md`.

I **did not** add any new tests; any suggestions on what to test on new apps would be appreciated!

image

image

@jnmarcus @adborden

### CODE REVIEW

File: disclosure_backend/management/commands/loadalldata.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import glob
+import os
+from django.conf import settings
+from django.core.management import call_command
+from calaccess_raw.management.commands import CalAccessCommand
+
+
+class Command(CalAccessCommand):
+    help = 'Generate documentation for raw CAL-ACCESS database models'
```
Comment (tdooner) [inline disclosure_backend/management/commands/loadalldata.py:9]: Either this comment is wrong, or the command is named poorly. (I would think `loadalldata` would be the combination of `downloadnetfilerawdata` and `downloadcalaccessrawdata`).

Comment (bcipolli) [inline disclosure_backend/management/commands/loadalldata.py:9]: Copy-paste error :) And actually, this is supposed to be an extension of Django's `loaddata` command, which loads a specified fixture. So hmmm, maybe I need a different name.

File: fips/views.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    Search for a location.
+    ---
+    parameters:
+      - name: q
+        description: The user's search query
+        type: string
+        paramType: query
+    """
+    query = request.query_params.get('q', '').lower()
+    query_set = FipsEntity.objects.filter(name__icontains=query)
```
Comment (tdooner) [inline fips/views.py:19]: Thoughts on `_icontains` vs. `_istartswith`? Probably most people won't try to find San Diego by starting with Diego. I think.

File: disclosure_backend/management/commands/loadalldata.py (hunk 1 above)
Comment (bcipolli) [inline disclosure_backend/management/commands/loadalldata.py:9]: (nice catch!)

**Comment by tdooner** [comment]
Looks promising. I'm interested in your thoughts on:

1) Is using `rest_framework` too heavy-handed?
2) FIPS codes strike me as somewhat of a rabbit hole: although useful to describe topology, they don't really get us closer to understanding the data because I don't believe any data actually is matchable to a FIPS code. Thus it's entirely a construct we're using to describe jurisdictions, which perhaps we could do more simply in another way. But to the extent that we will need to give each jurisdiction some ID eventually, I suppose the FIPS codes are a good place to start.

**Comment by bcipolli** [comment]
> 1) Is using rest_framework too heavy-handed?

I don't think so, I think it's perfect. I think it allows auto-discovery of methods and human-readable documentation that we wouldn't get by exposing APIs as raw XML. I like it a lot! Is there a similar option that's less heavy?

> 2) FIPS codes strike me as ... entirely a construct we're using to describe jurisdictions, which perhaps we could do more simply in another way.

Agreed; I wasn't sure what to call the model. I do think it's the right model, but seems like the wrong name. How about `Jurisdiction`? The `fips_id` is just the primary key for the Jurisdiction.

File: fips/models.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+from __future__ import unicode_literals
+from django.db import models
+from django.utils.encoding import python_2_unicode_compatible
+from calaccess_raw.models.base import CalAccessBaseModel
+
+
+@python_2_unicode_compatible
+class FipsEntity(CalAccessBaseModel):
```
Comment (adborden) [inline fips/models.py:8]: Not sure it's necessary to extend from `calAccessBaseModel`, the primary reason to do that would be if we're bulk loading from csv. Not sure if theres any other reason.

File: fips/serializers.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
+from rest_framework import serializers
+
+
+class FipsSerializer(serializers.Serializer):
+    name = serializers.CharField(max_length=1024)
+    type = serializers.CharField(max_length=2)
+    fips_id = serializers.IntegerField()
```
Comment (adborden) [inline fips/serializers.py:7]: Is this necesary? i thought that without specifying anything, the Serializer would default to exposing all fields on the model.

File: fips/views.py (hunk 2 above)
Comment (adborden) [inline fips/views.py:19]: Let's leave it as icontains for now. Would be good to call out for the user testing though cc @sseiter

File: fips/views.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+
+@api_view(['GET'])
+def search_view(request):
+    """
+    Search for a location.
+    ---
+    parameters:
+      - name: q
+        description: The user's search query
```
Comment (adborden) [inline fips/views.py:14]: This get's published in the API docs, so it's worth going the extra mile to make sure these are descriptive. How about:

> The user's search query against jurisdiction names.

**Comment by adborden** [comment]
This looks great! I like having small apps, even if they only contain a single endpoint and model.

I haven't used rest framework before, but I like what I see so far.

As for @tdooner's comment on the FIPS Id as a jurisdiction's primary key, I'm not sure where that request came from cc @jnmarcus @chellrocks. I know at one point we were talking about using the [Primary statistical area](https://en.wikipedia.org/wiki/United_States_primary_statistical_area) because it would help us identify what metro areas are related. i.e. We want to match money coming into Oakland from inside the Bay Area vs outside rather than simply within Oakland or not within Oakland. But I believe FIPS Ids get us this as well. Plus, we already have a map of [zipcodes to PSAs](http://admin.caciviclab.org/admin/zipcode_metro/zipcodemetro/) which was another requirement.

**Comment by bcipolli** [comment]
Thanks for the comments @adborden !

Looks to me that the `fipsEntity` or `Jurisdiction` is essentially a duplicate to the `Locality` model in the `ballot` app (https://github.com/caciviclab/disclosure-backend/blob/master/ballot/models.py#L35). I realize I need to think through the database schema as a whole, and front-end needs, to think how best to resolve that (e.g. modifying the `ballot` app rather than making this new app).

**Comment by bcipolli** [comment]
To merge the schemas re: above, and enable other feeds as requested, I have begun (on this branch) to prototype the schema described here: https://github.com/caciviclab/caciviclab.github.io/wiki/Data-relationships---Database-schema

**Comment by bcipolli** [comment]
This can be closed; it's outlived it's utility. This was mostly a demo for @jnmarcus ; we've moved on to bigger and better things :)

**Pull request closed without merging**